cpp/src/arrow/ipc/reader.h

```diff
@@ -169,6 +170,17 @@ class ARROW_EXPORT RecordBatchFileReader {
 
   /// \brief Return current read statistics
   virtual ReadStats stats() const = 0;
+
+  /// \brief Get a reentrant generator of record batches.
+  ///
+  /// This RecordBatchFileReader must outlive the generator.
```

Comment:
Given that we're talking about an asynchronous primitive, forcing the caller to deal with lifetime issues is a bit unfriendly IMHO. It may be simpler to make `RecordBatchFileReader` inherit from `enable_shared_from_this` and capture a strong reference inside the generator.

Reply:
Thanks - this was actually left over from an earlier approach. I've removed the note and adjusted the test to ensure the reader gets dropped before the generator.